src/librawspeed/decoders/ArwDecoder.cpp

```diff
@@ -258,6 +264,78 @@ void ArwDecoder::DecodeUncompressed(const TiffIFD* raw) const {
     u.decodeRawUnpacked<16, Endianness::little>(width, height);
 }
 
+void ArwDecoder::DecodeLJpeg(const TiffIFD* raw) const {
+  uint32_t width = raw->getEntry(TiffTag::IMAGEWIDTH)->getU32();
+  uint32_t height = raw->getEntry(TiffTag::IMAGELENGTH)->getU32();
+  uint32_t bitPerPixel = raw->getEntry(TiffTag::BITSPERSAMPLE)->getU32();
+
+  switch (bitPerPixel) {
+  case 8:
+  case 12:
+  case 14:
+    break;
+  default:
+    ThrowRDE("Unexpected bits per pixel: %u", bitPerPixel);
+  }
+
+  if (width == 0 || height == 0 || height % 2 != 0 || width > 9600 ||
+      height > 6376)
+    ThrowRDE("Unexpected image dimensions found: (%u; %u)", width, height);
+
+  mRaw->dim = iPoint2D(width, height);
+
+
+  const uint32_t tilew = raw->getEntry(TiffTag::TILEWIDTH)->getU32();
+  const uint32_t tileh = raw->getEntry(TiffTag::TILELENGTH)->getU32();
+
+  if (!(tilew > 0 && tileh > 0))
+    ThrowRDE("Invalid tile size: (%u, %u)", tilew, tileh);
+
+  assert(tilew > 0);
+  const uint32_t tilesX = roundUpDivision(mRaw->dim.x, tilew);
+  if (!tilesX)
+    ThrowRDE("Zero tiles horizontally");
+
+  assert(tileh > 0);
+  const uint32_t tilesY = roundUpDivision(mRaw->dim.y, tileh);
+  if (!tilesY)
+    ThrowRDE("Zero tiles vertically");
+
+  const TiffEntry* offsets = raw->getEntry(TiffTag::TILEOFFSETS);
+  const TiffEntry* counts = raw->getEntry(TiffTag::TILEBYTECOUNTS);
+  if (offsets->count != counts->count) {
+    ThrowRDE("Tile count mismatch: offsets:%u count:%u", offsets->count,
+             counts->count);
+  }
+
+  // tilesX * tilesY may overflow, but division is fine, so let's do that.
+  if ((offsets->count / tilesX != tilesY || (offsets->count % tilesX != 0)) ||
+      (offsets->count / tilesY != tilesX || (offsets->count % tilesY != 0))) {
+    ThrowRDE("Tile X/Y count mismatch: total:%u X:%u, Y:%u", offsets->count,
+             tilesX, tilesY);
+  }
+
+  mRaw->createData();
+  for (uint32_t tile = 0; tile < offsets->count; tile++) {
+    const uint32_t tileX = tile % tilesX;
+    const uint32_t tileY = tile / tilesX;
+    const uint32_t offset = offsets->getU32(tile);
+    const uint32_t length = counts->getU32(tile);
+    LJpegDecompressor decompressor(
+        ByteStream(
+            DataBuffer(mFile.getSubView(offset, length), Endianness::little)),
+        mRaw);
+    decompressor.decode(tileX * tilew, tileY * tileh, tilew, tileh, false,
+                        true);
+  }
+
+  const TiffEntry* origin_entry = raw->getEntry(TiffTag::DEFAULTCROPORIGIN);
+  const TiffEntry* size_entry = raw->getEntry(TiffTag::DEFAULTCROPSIZE);
+  iRectangle2D crop(origin_entry->getU32(0), origin_entry->getU32(1),
+                    size_entry->getU32(0), size_entry->getU32(1));
+  mRaw->subFrame(crop);
```

Comment:
I don't think we use the file crop in any other modes (uncompressed and lossy), but rather leave it to the settings from XML? This way, there will be a difference between the modes.

There is a small problem there, as width/height for lossless is wrongly specified by Sony as integer number of tiles, and it didn't have to be... [One way to handle this was to have always an absolute positive crop in the XML for all modes, but it was rejected...](https://github.com/darktable-org/rawspeed/pull/274)

Reply:
If I remove it then the original under history has black bars on the bottom right, although it is cropped out later in the default module history.

Reply:
Sure, I didn't mean remove the crop completely. But we'll have to find a way to provide the same crops as uncompressed/lossy mode (defined in data/cameras.xml), and that works for APC-C/S35 files as well.

Reply:
@artemist I think this tag has the information of the uncropped size that you actually want, e.g. for the 7M4:
```
Exif.SubImage1.0x7038                          2  7032 4688
```

Reply:
> If I remove it then the original under history has black bars on the bottom right, although it is cropped out later in the default module history.

This is the correct behaviour.

> <...>

Since we don't use the camera-specified crop,
i don't really see why we care what the EXIF says,
we are just going to re-use the existing cameras.xml-specified crop in that mode too,
unless we need a different one, like some Fuji's need.

Reply:
> i don't really see why we care what the EXIF says,

Because of this new lossless mode and tile size rounding up nonsense. This is actually the one thing that correlates it to the traditional modes and cameras.xml (unless you want to bloat cameras.xml with more entries, which I don't think is a good idea).

I'll provide an example if it's not clear

Reply:
@LebedevRI Ok, let's take the 7M4 here - in the traditional uncompressed and lossy modes it has 
```
Exif.SubImage1.ImageWidth                     7040
Exif.SubImage1.ImageLength                    4688
```
and in (tweaked) cameras.xml we put `<Crop x="0" y="0" width="-8" height="0"/>`, so the resulting raw image size is `7032 4688`.

In the new lossless mode it has this nonsensical round up to integer multiple of 512x512 tiles:
```
Exif.SubImage1.ImageWidth                     7168
Exif.SubImage1.ImageLength                    5120
```
and the new tag
```
Exif.SubImage1.0x7038                             7032 4688
```

See the pattern here?

So options for the lossless mode implementation to give us the same output size as uncompressed are 1) bloat cameras.xml with more entries just so we can have different crop numbers (and not just one; we'll need extra for APS-C outputs as remainder after rounding up is different of course) or 2) just use the bloody 0x7038 tag as is because it is exactly what it means - **_maximum usable raw area._**

I have verified all the ILCE-1/7M4/7SM3/7RM5 samples on RPU (dump using LibRaw's `unprocessed_raw` which also works for lossless) and everything matches up, and it also lines up perfectly for the APS-C mode of 7RM5 and cameras.xml crop for the corresponding uncompressed.

> i don't really see why we care what the EXIF says.

And a final thing - **_this is not "crop according to EXIF",_** as we have just established. That would in fact be further cropping according to
```
Exif.SubImage1.DefaultCropOrigin              12 8
Exif.SubImage1.DefaultCropSize                7008 4672
```
which is what @artemist is doing now (and matching OOC JPEG size). I agree, _this_ is not what we generally want or care about.